### PR TITLE
Add support of h3ulcb and salvator xs m3 arm64 boards

### PR DESCRIPTION
**Pull request by xakep-amatop**
Add generic Renesas functions to CPG for getting/setting rates of clocks. Add associated tables of clocks and API for working with these tables.

Add MMIO mapping for PFC, UART and CPG Renesas drivers in order to avoid mappings inside `mmu_regions.c` file(s).

Added files for supporting  ARM64 Renesas `r8a77951`/`r8a77961` SoCs:
    * add dtsi for `r8a77951`/`r8a77961` soc;
    * files for support PFC;

Add basic support of `rcar_salvator_xs_m3` and `rcar_h3ulcb_ca57` ARM64 boards.

### CODE REVIEW

**Review by gmarull — DISMISSED** [review]

**Comment by aaillet** [comment]
Hi @xakep-amatop ,
Before reviewing this work, I would like to know which use cases you are targeting by supporting Cortex-A57 core of these boards in Zephyr.
These boards are equipped with real-time purpose Cortex-R7 cores which we are already supporting (only H3 based boards atm).
A57s and R7 cores may even be able to communicate together through shared msg queues (work is in progress).

I would like to know your though (and even other community members though) about this.

In the case we want to merge this work, I think it would be nice to unify as much as possible our work from the beginning to create a clean working tree containing both arm and arm64 part of these SoCs support.

**Comment by lorc** [comment]
Hi @aaillet,

Let me jump in with some clarifications.

> Before reviewing this work, I would like to know which use cases you are targeting by supporting Cortex-A57 core of these boards in Zephyr.

We are working on virtualization on Rcar Gen3/Gen4 platform. If you are interested, you can check out our repo: https://github.com/xen-troops/meta-xt-prod-devel-rcar 

Here is some background: we are running Xen Hypervisor. This is Type-1 hypervisor, which means that it boots before any OS. After boot Xen automatically creates very first virtual machine, which is known as Dom0. This is a bit special VM, designated as  "control" or "privileged" domain, because it is capable of controlling hypervisor and creating other VMs (or "domains", as Xen call them). In most cases Dom0 is a typical Linux distribution with some additional Xen-related tools. This is perfectly fine for generic server use cases, but is not suitable for automotive ones. 

We are planning Xen to be safety-certified, but we need safety-certified Dom0 as well, because Dom0 can control hypervisor behaviour. As it is almost impossible to certify Linux kernel, we need something more specialized and lightweight, which is Zephyr, of course :)

Actually, we already have "Zephyr as Dom0" solution. It is not ready for production yet, but if anyone want to experiment with it, we can provide links to the repos. We are planning to upstream all relevant changes and this PR is the first in series.

So yeah, to run Zephyr as Xen's Dom0, we need Zephyr to be able to run on CortexA cores on RCar Gen3 first. This is why you are seeing this PR.

File: boards/arm64/renesas_rcar/gen3/rcar_salvator_xs_m3_2x4g/doc/rcar_salvator_xs_m3_2x4g.rst
Comment (SgrrZhf) [inline boards/arm64/renesas_rcar/gen3/rcar_salvator_xs_m3_2x4g/doc/rcar_salvator_xs_m3_2x4g.rst:33]: Missing a blank line?

https://builds.zephyrproject.io/zephyr/pr/57172/docs/boards/arm64/renesas_rcar/gen3/rcar_salvator_xs_m3_2x4g/doc/rcar_salvator_xs_m3_2x4g.html

File: boards/arm64/renesas_rcar/Kconfig.board
Comment (aaillet) [inline boards/arm64/renesas_rcar/Kconfig.board:1]: boards file tree is usually not using brand/family folders.
Please get boards folder to the same tree level = boards/arm64/rcar_h3ulcb_ca57

File: dts/arm64/renesas/gen3_common/r8a7795_r8a7796_common.dtsi
Comment (aaillet) [inline dts/arm64/renesas/gen3_common/r8a7795_r8a7796_common.dtsi:1]: Please use same file tree as for arm32 part of renesas socs for clarity purpose.
dts/arm64/renesas/gen3 with
  - rcar_gen3_ca57.dtsi (common)
  - r8a77951.dtsi (r8a77950 SoC is deprecated, no need to support it)
  - r8a7796.dtsi (M3-W + M3-N)

File: soc/arm64/renesas_rcar/gen3/Kconfig.defconfig.r8a7795
Comment (aaillet) [inline soc/arm64/renesas_rcar/gen3/Kconfig.defconfig.r8a7795:1]: please rename to Kconfig.defconfig.r8a77951, we dont support deprecated r8a77950 SoC

File: soc/arm64/renesas_rcar/gen3/pfc_r8a7795.c
Comment (aaillet) [inline soc/arm64/renesas_rcar/gen3/pfc_r8a7795.c:1]: please rename to soc/arm64/renesas_rcar/gen3/pfc_r8a77951.c

File: boards/arm64/renesas_rcar/gen3/common_salvator_xs_h3ulcb.dts
Comment (aaillet) [inline boards/arm64/renesas_rcar/gen3/common_salvator_xs_h3ulcb.dts:1]: It's currently not common to deploy common files in the boards file tree, please put back these fields in each board dts.

**Review by aaillet — DISMISSED** [review]
Hi, I did a first short review on file tree, please edit it to match what have been done on CR7 side and common Zephyr practice.

Concerning using MMIO mapping in several drivers, are these changes mandatory for the A57 to use these drivers ? If not, please remove these commits from the PR and do another dedicated PR to use MMIO mapping on all r_car drivers !

We should check if we can use one and only `pfc_r8a77951.c` and so on.

Also, for the Zephyr bot to be able to assign the right people when creating Pull requests, separate label should be created for Renesas R-Car arm and Renesas R-Car arm64. Maybe you can add a commit editing `MAINTAINERS.yml`.

I haven't had time to review the rest of the changes yet and I won't be around for two weeks, sorry for that, I'll resume the review then.

**Comment by xakep-amatop** [comment]
> Hi, I did a first short review on file tree, please edit it to match what have been done on CR7 side and common Zephyr practice.

thanks, fixed!

> Concerning using MMIO mapping in several drivers, are these changes mandatory for the A57 to use these drivers ? If not, please remove these commits from the PR and do another dedicated PR to use MMIO mapping on all r_car drivers !

most of changes inside drivers are necessary in order to get working UART and to test the code, so I've put drivers related changes to this PR

> We should check if we can use one and only pfc_r8a77951.c and so on.

do you mean to share this file between arm32 & arm64 SoCs or only between r8a77951 & r8a7796 for arm64?

> Also, for the Zephyr bot to be able to assign the right people when creating Pull requests, separate label should be created for Renesas R-Car arm and Renesas R-Car arm64. Maybe you can add a commit editing MAINTAINERS.yml.

Do want to become a maintainer for Renesas R-Car arm64 too? I'm afraid that I have not enough experience for this role. Maybe you know somebody who wants and can be a maintainer?

**Review by harunobukurokawa — COMMENTED** [review]
Hi, 
Please confirm the revision of M3. These are difference between "r8a77960" and "r8a77961".

File: dts/arm64/renesas/r8a7796.dtsi
Comment (xakep-amatop) [reply dts/arm64/renesas/r8a7796.dtsi:1]: I've used Salvator XS M3 ver 3.0 for tests

according to this page https://elinux.org/R-Car/Boards/Salvator-XS:
both r8a77960 and r8a77961 for M3:
```
 * DTB: arch/arm64/boot/dts/renesas/r8a77960-salvator-xs.dtb (for R-Car M3-W)
 * DTB: arch/arm64/boot/dts/renesas/r8a77961-salvator-xs.dtb (for R-Car M3-W+)
 * DTB: arch/arm64/boot/dts/renesas/r8a77965-salvator-xs.dtb (for R-Car M3-N)
```

u-boot on test board reports the next info:
```
CPU: Renesas Electronics R8A7796 rev 3.0
Model: Renesas Salvator-X board based on r8a77960
DRAM:  7.9 GiB
```

but sticker on board says that this is XS board and BL2 approves this info:
```
NOTICE:  BL2: R-Car Gen3 Initial Program Loader(CA57) Rev.3.0.1
NOTICE:  BL2: PRR is R-Car M3 Ver.3.0
NOTICE:  BL2: Board is Salvator-XS Rev.1.0
```

Comment (harunobukurokawa) [reply dts/arm64/renesas/r8a7796.dtsi:1]: Thank you for your information. 
It is "r8a77961" because you use M3 ver 3.0. The description from u-boot is not correct.

**Comment by xakep-amatop** [comment]
fixed naming of files used for documentation

File: dts/arm64/renesas/r8a7796.dtsi
Comment (xakep-amatop) [reply dts/arm64/renesas/r8a7796.dtsi]: used `r8a77961` name instead of `r8a7796`/`r8a77960`

**Review by carlocaione — DISMISSED** [review]

**Review by harunobukurokawa — DISMISSED** [review]

**Comment by carlescufi** [comment]
@xakep-amatop please rebase
@gmarull @aaillet please re-review

**Comment by carlescufi** [comment]
> > @xakep-amatop please rebase @gmarull @aaillet please re-review
> 
> @carlescufi thank you, I've fixed merge conflicts!

@xakep-amatop thanks, please take a look at the failing compliance check.

**Comment by xakep-amatop** [comment]
> > > @xakep-amatop please rebase @gmarull @aaillet please re-review
> > 
> > 
> > @carlescufi thank you, I've fixed merge conflicts!
> 
> @xakep-amatop thanks, please take a look at the failing compliance check.

@carlescufi as far as I understand using of `typedef`s for `pinctrl_soc_pin_t` it is a common approach, look at how it was made for other SoCs (part of output):
```
###### grep -rnI pinctrl_soc_pin_t

arm64/nxp_imx/mimx9/pinctrl_soc.h:52:typedef struct pinctrl_soc_pin pinctrl_soc_pin_t;
arm64/nxp_imx/mimx8m/pinctrl_soc.h:52:typedef struct pinctrl_soc_pin pinctrl_soc_pin_t;
arm64/renesas_rcar/gen4/pinctrl_soc.h:41:} pinctrl_soc_pin_t;
riscv/esp32c3/pinctrl_soc.h:28:} pinctrl_soc_pin_t;
riscv/openisa_rv32m1/pinctrl_soc.h:23:typedef uint32_t pinctrl_soc_pin_t;
riscv/riscv-privilege/telink_b91/pinctrl_soc.h:17:typedef uint32_t pinctrl_soc_pin_t;
riscv/riscv-privilege/sifive-freedom/pinctrl_soc.h:12:typedef struct pinctrl_soc_pin_t {
riscv/riscv-privilege/sifive-freedom/pinctrl_soc.h:15:} pinctrl_soc_pin_t;
riscv/riscv-ite/common/pinctrl_soc.h:33:typedef struct pinctrl_soc_pin pinctrl_soc_pin_t;
xtensa/esp32s2/pinctrl_soc.h:28:} pinctrl_soc_pin_t;
xtensa/esp32/pinctrl_soc.h:28:} pinctrl_soc_pin_t;
arm/nuvoton_npcx/common/pinctrl_soc.h:125:typedef struct npcx_pinctrl pinctrl_soc_pin_t;
arm/rpi_pico/rp2/pinctrl_soc.h:34:typedef struct rpi_pinctrl_soc_pin pinctrl_soc_pin_t;
arm/renesas_smartbond/da1469x/pinctrl_soc.h:20:typedef struct smartbond_pinctrl_soc_pin pinctrl_soc_pin_t;
```

so, should the issue be fixed?

**Review by carlocaione — DISMISSED** [review]
LGTM for the architectural part

**Comment by xakep-amatop** [comment]
Hi, @aaillet
could you please re-review changes and answer the question ?

**Comment by xakep-amatop** [comment]
fixed merging conflict

**Comment by xakep-amatop** [comment]
fixed twister `tests/subsys/mgmt/mcumgr/smp_version` test for a new platforms

**Comment by aaillet** [comment]
> Hi, @aaillet could you please re-review changes and answer the question ?

Working on another review !

Concerning questions, it's about maintainer ?
I talked with my internal Renesas contact and we are not sure yet of which team should maintain this part !
I don't think we have the time at Iot.bzh to maintain and test correctly arm64 part of these SoCs.
Discussions are in progress.

**Comment by xakep-amatop** [comment]
> > Hi, @aaillet could you please re-review changes and answer the question ?
> 
> Working on a review !
> 
> Concerning questions, it's about maintainer ? I talked with my internal Renesas contact and we are not sure yet of which team should maintain this part ! I don't think we have the time at Iot.bzh to maintain and test correctly arm64 part of these SoCs. Discussions are in progress.

got it, thank you

**Comment by lorc** [comment]
Hi @aaillet,

I can maintain this part of the code. Our company (EPAM Systems) is working closely with Renesas also.  I am already maintainer of Renesas platform in OP-TEE: https://github.com/OP-TEE/optee_os/blob/master/MAINTAINERS#L220 and collaborator of Xen-related stuff in Zephyr: https://github.com/zephyrproject-rtos/zephyr/blob/main/MAINTAINERS.yml#L1461
 
So I am ready to act as maintainer/collaborator for GEN3/aarch64 part as well.

**Review by carlocaione — DISMISSED** [review]
LGTM for the arch part

**Comment by xakep-amatop** [comment]
Hi, @aaillet 

I would really apreciate if you could review my changes

**Comment by lorc** [comment]
Hi @aaillet,

If you are okay with having me as maintainer, could you please point @xakep-amatop where he should put me in the `MAINTAINERS.yml`  ? Should I be in the `maintainers` or `collaborators` section?

**Comment by aaillet** [comment]
> Hi @aaillet,
> 
> If you are okay with having me as maintainer, could you please point @xakep-amatop where he should put me in the `MAINTAINERS.yml` ? Should I be in the `maintainers` or `collaborators` section?

Hi, there is two possible solution:
- @xakep-amatop is creating a new "renesas r-car arm64" area to the maintainer file and add you as maintainer
- I create another PR when this one will be merged to do it my self

Creating a new "area" should be done alongside the add of a new github label for the bot to assign. The new maintainer will also be added to zephyr github teams to get corresponding rights on the project.



> Hi, @aaillet
> 
> I would really apreciate if you could review my changes

Sorry for the delay, I need to find time to do it asap, I'm starting to test new drivers version on our arm32 targets today to validate functioning.

**Review by de-nordic — DISMISSED** [review]

**Review by nordicjm — DISMISSED** [review]
MCUmgr change OK

**Comment by aaillet** [comment]
I validated that these changes didn't broke R7 (arm32) support of H3.
I'm deeply reviewing clock control changes, I'll let you know.

Can you tell me what do you want to do concerning maintainer assignment ?

> Hi, there is two possible solution:
> 
> * @xakep-amatop is creating a new "renesas r-car arm64" area to the maintainer file and add you as maintainer
> * I create another PR when this one will be merged to do it my self (I have some update to my area to do, so I can do it at the same time)

**Comment by xakep-amatop** [comment]
rebased onto the latest changes + added `Renesas R-Car ARM64` entry to the maintainers list ([last commit](https://github.com/zephyrproject-rtos/zephyr/pull/57172/commits/01f74f9eea8b08436a1f84a9ce4fbbc24d793b5e))

**Review by carlocaione — APPROVED** [review]
For the arch part

**Review by lorc — APPROVED** [review]
For adding me as a maintainer

**Review by de-nordic — APPROVED** [review]
MCUmgr is OK.

**Comment by carlescufi** [comment]
Compliance failure is the usual pin control typedef.

**Comment by aaillet** [comment]
In fact PR wasn't ready to be merged, I was working on the clock changes review as it is containing errors and kept my "change needed" while doing it.
Sorry I took that much time, I will then address these errors in a new PR !

**Comment by carlocaione** [comment]
> In fact PR wasn't ready to be merged

Next time please keep it in draft state until ready to be merged.

**Comment by aaillet** [comment]
> > In fact PR wasn't ready to be merged
> 
> Next time please keep it in draft state until ready to be merged.

I'm not PR author, was I able to do that ?
I wasn't aware, that's why I used "changes needed" to block the PR

**Comment by xakep-amatop** [comment]
> > I validated that these changes didn't broke R7 (arm32) support of H3. I'm deeply reviewing clock control changes, I'll let you know.
> 
> I said that I was working on the clock control review (but not updated my status) as maintainer, but anyway, I'll fix it on another PR !

Feel free to write comments about any issues you've found in the CPG driver here. If you'd like, I can address them and open a new pull request.

**Pull request merged**
